File: app/src/main/java/com/bis5/fitjourney/viewmodels/WorkoutViewModelFactory.java
package com.bis5.fitjourney.viewmodels;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.bis5.fitjourney.models.ExerciseDao;
import com.bis5.fitjourney.models.SetLogDao;
import com.bis5.fitjourney.models.WorkoutDao;
import com.bis5.fitjourney.models.WorkoutLogDao;

public class WorkoutViewModelFactory implements ViewModelProvider.Factory {

    private final WorkoutDao workoutDao;
    private final ExerciseDao exerciseDao;
    private final WorkoutLogDao workoutLogDao;
    private final SetLogDao setLogDao;

    public WorkoutViewModelFactory(WorkoutDao workoutDao, ExerciseDao exerciseDao, WorkoutLogDao workoutLogDao, SetLogDao setLogDao) {
        this.workoutDao = workoutDao;
        this.exerciseDao = exerciseDao;
        this.workoutLogDao = workoutLogDao;
        this.setLogDao = setLogDao;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(WorkoutViewModel.class)) {
            // Use direct instantiation instead of reflection. It is safer and simpler.
            return (T) new WorkoutViewModel(workoutDao, exerciseDao, workoutLogDao, setLogDao);
        }
        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }
}

File: app/src/main/java/com/bis5/fitjourney/viewmodels/WorkoutViewModelFactory.kt
package com.bis5.fitjourney.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.bis5.fitjourney.models.ExerciseDao
import com.bis5.fitjourney.models.SetLogDao
import com.bis5.fitjourney.models.WorkoutDao
import com.bis5.fitjourney.models.WorkoutLogDao

class WorkoutViewModelFactory(private val workoutDao: WorkoutDao, private val exerciseDao: ExerciseDao, private val workoutLogDao: WorkoutLogDao, private val setLogDao: SetLogDao) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(WorkoutViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return WorkoutViewModel(workoutDao, exerciseDao, workoutLogDao, setLogDao) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}
